src/index.ts

```diff
@@ -1,173 +1,231 @@
+  tooltipElement.setAttribute('id', options.id);
+  tooltipElement.classList.add('vg-tooltip');
+
+  document.querySelector('body')!.appendChild(tooltipElement);
 }
 
-export function vegaLite(vgView: VgView, vlSpec: TopLevelSpec, options: Option = {showAllFields: true, isComposition: false}) {
-  options = supplementOptions(copyOptions(options), vlSpec);
-  start(vgView, copyOptions(options));
+/**
+ * Format the value to be shown in the toolip.
+ *
+ * @param value The value to show in the tooltip.
+ */
+function formatValue(value: any): string {
+  if (isArray(value)) {
+    return `[${value.map(v => (isString(v) ? v : stringify(v))).join(', ')}]`;
+  }
 
-  return {
-    destroy: function () {
-      // remove event listeners
-      vgView.removeEventListener('mouseover.tooltipInit');
-      vgView.removeEventListener('mousemove.tooltipUpdate');
-      vgView.removeEventListener('mouseout.tooltipRemove');
+  if (isString(value)) {
+    return value;
+  }
 
-      cancelPromise(); // clear tooltip promise
-    }
-  };
-}
+  if (isObject(value)) {
+    let content = '';
 
-function start(vgView: VgView, options: Option) {
-  // initialize tooltip with item data and options on mouse over
-  vgView.addEventListener('mouseover.tooltipInit', function (event: MouseEvent, item: Scenegraph) {
-    if (shouldShowTooltip(item)) {
-      // clear existing promise because mouse can only point at one thing at a time
-      cancelPromise();
-
-      // make a new promise with time delay for tooltip
-      tooltipPromise = window.setTimeout(function () {
-        init(event, item, options);
-      }, options.delay || DELAY);
-    }
-  });
+    const { title, ...rest } = value as any;
 
-  // update tooltip position on mouse move
-  // (important for large marks e.g. bars)
-  vgView.addEventListener('mousemove.tooltipUpdate', function (event: MouseEvent, item: Scenegraph) {
-    if (shouldShowTooltip(item) && tooltipActive) {
-      update(event, item, options);
+    if (title) {
+      content += `<h2>${title}</h2>`;
     }
-  });
-
-  // clear tooltip on mouse out
-  vgView.addEventListener('mouseout.tooltipRemove', function (event: MouseEvent, item: Scenegraph) {
-    if (shouldShowTooltip(item)) {
-      cancelPromise();
 
-      if (tooltipActive) {
-        clear(event, item, options);
+    content += '<table>';
+    for (const key of Object.keys(rest)) {
+      let val = (rest as any)[key];
+      if (isObject(val)) {
+        val = stringify(val);
       }
+
+      content += `<tr><td class="key">${key}:</td><td class="value">${val}</td></tr>`;
     }
-  });
-}
+    content += `</table>`;
+
+    return content;
+  }
 
-/* Cancel tooltip promise */
-function cancelPromise() {
-  /* We don't check if tooltipPromise is valid because passing
-   an invalid ID to clearTimeout does not have any effect
-   (and doesn't throw an exception). */
-  window.clearTimeout(tooltipPromise);
-  tooltipPromise = undefined;
+  return `${value}`;
```

Comment:
`String(value)`  ?  Otherwise you are simply using a complex (slow) mechanism to convert value to a string.